rpc/codec/server.go

```diff
@@ -37,6 +38,9 @@ type serverCodec struct {
 
 	methods []string
 
+	// Authentication hook. May be nil in tests.
+	authenticationHook func(string) error
```

Comment:
The hook doesn't need to be authentication specific. Make the function take a `proto.Message` and have the proto perform a type assertion or type switch to extract the user.